clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -903,6 +904,54 @@ private void updateBytesBuffered(boolean madeProgress) {
 
     @Override
     public long transferFrom(FileChannel fileChannel, long position, long count) throws IOException {
-        return fileChannel.transferTo(position, count, this);
+        if (state == State.CLOSING)
+            throw closingException();
+        if (state != State.READY)
+            return 0;
+
+        if (!flush(netWriteBuffer))
+            return 0;
+
+        long channelSize = fileChannel.size();
+        if (position > channelSize)
+            return 0;
+        int bytesToWrite = (int) Math.min(Math.min(count, channelSize - position), Integer.MAX_VALUE);
+
+        if (fileChannelBuffer == null) {
+            // same size as `netWriteBuffer`
+            int transferSize = 16384;
+            // Allocate a direct buffer to avoid one heap to heap buffer copy. SSLEngine copies the source
+            // buffer (fileChannelBuffer) to the destination buffer (netWriteBuffer) and then encrypts in-place.
+            // FileChannel.read() to a heap buffer requires a copy from a direct buffer to a heap buffer, which is not
+            // useful here.
+            fileChannelBuffer = ByteBuffer.allocateDirect(transferSize);
+        } else {
+            fileChannelBuffer.clear();
+        }
+
+        int bytesWritten = 0;
+        long pos = position;
+        try {
+            while (bytesWritten < bytesToWrite) {
+                int bytesRemaining = bytesToWrite - bytesWritten;
+                if (bytesRemaining < fileChannelBuffer.limit())
+                    fileChannelBuffer.limit(bytesRemaining);
+                int bytesRead = fileChannel.read(fileChannelBuffer, pos);
+                if (bytesRead <= 0)
+                    break;
+                fileChannelBuffer.flip();
+                int networkBytesWritten = write(fileChannelBuffer);
+                bytesWritten += networkBytesWritten;
+                if (networkBytesWritten != bytesRead)
+                    break;
+                pos += networkBytesWritten;
+                fileChannelBuffer.clear();
+            }
+            return bytesWritten;
+        } catch (IOException x) {
+            if (bytesWritten > 0)
+                return bytesWritten;
```

Comment:
This maintains existing behavior, but seems weird. Shouldn't we be propagating the exception? For example, the disk could be bad.